### PR TITLE
Add kerberos example

### PR DESCRIPTION
**Pull request by adfoster-r7**
I thought it'd be useful to put up a PR to for the Kerberos example models that I'm working through, similar to the complex SNMP example - https://github.com/sdaubert/rasn1/wiki/Complex-example%3A-SNMP

It's still in draft for now, as there's some open PRs/issues/models to work through and add still

### CODE REVIEW

File: spec/model_spec.rb [hunk 1]
```diff
@@ -48,6 +48,11 @@ class ExplicitTaggedSeq < Model
              content: [integer(:id), integer(:extern_id)]
   end
 
+  class ModelExplicitBitString < Model
```
Comment (adfoster-r7) [inline spec/model_spec.rb:51]: Separate PR for this https://github.com/sdaubert/rasn1/pull/15

File: lib/rasn1/types/base.rb [hunk 2]
```diff
@@ -360,6 +360,7 @@ def build
           if explicit?
             v = explicit_type
             v.value = @value
+            v.bit_length = @bit_length if @bit_length
```
Comment (adfoster-r7) [inline lib/rasn1/types/base.rb:363]: Separate PR for this https://github.com/sdaubert/rasn1/pull/15/files

File: Gemfile [hunk 3]
```diff
@@ -8,7 +8,8 @@ gemspec
 gem 'bundler', '>=1.17', '<3'
 
 group :noci do
-  gem 'debase', '~>0.2'
+  # Broken on Ruby 3.0.2 - https://github.com/ruby-debug/debase/issues/98
```
Comment (adfoster-r7) [inline Gemfile:11]: Will revert this of course; Just putting up the example PR for now :+1:

File: lib/rasn1/model.rb [hunk 4]
```diff
@@ -588,10 +588,19 @@ def sequence_to_h(seq)
...
+               when Wrapper
+                 @elements.key(el.element)
                else
                  el.name
                end
-        [name, private_to_h(el)]
+
+        if name.nil?
+          # $stderr.puts "name is nil"
+          # name = @elements[:ticket].instance_variable_get(:@elements).key(el.element)
```
Comment (adfoster-r7) [inline lib/rasn1/model.rb:599]: Adding this line back in allows the tests to pass

Comment (adfoster-r7) [inline lib/rasn1/model.rb:599]: Adding this line back in allows the tests to pass; Might require additional changes to https://github.com/sdaubert/rasn1/pull/22

**Comment by sdaubert** [comment]
These specs may be merged, but hey have to use RASN1 custom/constrained types instead of defining methods in Model class.

**Comment by adfoster-r7** [comment]
> These specs may be merged, but hey have to use RASN1 custom/constrained types instead of defining methods in Model class.

I'd like to avoid mutating the RASN1 namespace as part of the test suite if possible, as it might introduce issues with non-deterministic test results depending on test order. A similar rationale exists for avoiding the pattern in end user's code, where I'd like to avoid the scenario where different protocols have their own definitions of what primitives are - such as Kerberos constraining 'microseconds' to `(0..999999)`, but another protocol being unbounded.

Would you be okay with using an inheritance approach instead?

```ruby
  class Int32 < RASN1::Types::Integer
    # ...
  end
 
  class KerberosModel < RASN1::Model
    def self.kerberos_uint32(name, options = {})
      custom_primitive_type_for(name, UInt32, options)
    end

    def self.kerberos_int32(name, options = {})
      custom_primitive_type_for(name, Int32, options)
    end

    # ...

    def self.custom_primitive_type_for(name, clazz, options = {})
      options.merge!(name: name)
      proc = proc do |opts|
        clazz.new(options.merge(opts))
      end
      @root = Elem.new(name, proc, nil)
    end

    private_class_method :custom_primitive_type_for
  end

  class EncryptedData < KerberosModel
    sequence :encrypted_data,
             content: [
               kerberos_int32(:etype, explicit: 0, constructed: true),
               kerberos_uint32(:kvno, explicit: 1, constructed: true, optional: true),
               octet_string(:cipher, explicit: 2, constructed: true)
             ]
  end
```

I've just pushed that approach as a commit now for the full example

**Comment by sdaubert** [comment]
> I'd like to avoid mutating the RASN1 namespace as part of the test suite if possible, as it might introduce issues with non-deterministic test results depending on test order. 

I agree you. I thought about this last days, and my current implementation is not so good...

For the implementation, i would want to stay with constrained types, but in the following ways:

```ruby
# First option
module MyModule
  # use an option (`in` here) to specify in which module the new type will be defined
  RASN1::Types.define_type('uint32', from: RASN1::Types::Integer, in: self) do |value|
    (value >= 0) && (value < 2**32)
  end

  class MyClass < RASN1::Model
    # ...
  end
end

# Second option
module MyModule
  # automagically define the new type in current module. I am not sure this is possible
  RASN1::Types.define_type('uint32', from: RASN1::Types::Integer) do |value|
    (value >= 0) && (value < 2**32)
  end

  class MyClass < RASN1::Model
    # ...
  end
end
```

**Comment by sdaubert** [comment]
See #23 about redefining `RASN1::Types.define_type`

**Comment by adfoster-r7** [comment]
Was there anything else required for this PR? Everything is green now from when it was last pushed

**Comment by sdaubert** [comment]
Please remove your change in Gemfile.
Your specific Kerberos types (Int32, UInt32 and Microseconds, and maybe others) should be defined using RASN1::Types.define_type. This ways you can easily add constraints.

**Comment by adfoster-r7** [comment]
Sorry for the delay; Updated to use define_type :+1:

**Pull request merged**